File: Safari-609.1.11.1/LayoutTests/http/tests/blink/sendbeacon/resources/check-beacon.php
<?php
require_once '../../../resources/portabilityLayer.php';

$beaconFilename = sys_get_temp_dir() . "/beacon" . (isset($_REQUEST['name']) ? $_REQUEST['name'] : "") . ".txt";

$max_attempts = 700;
$retries = isset($_REQUEST['retries']) ? (int)$_REQUEST['retries'] : $max_attempts;
while (!file_exists($beaconFilename) && $retries != 0) {
    usleep(10000);
    # file_exists() caches results, we want to invalidate the cache.
    clearstatcache();
    $retries--;
}

header('Content-Type: text/plain');
header('Access-Control-Allow-Origin: *');
if (file_exists($beaconFilename)) {
    $beaconFile = false;
    if (is_readable($beaconFilename)) {
        $beaconFile = fopen($beaconFilename, 'r');
    }
    if ($beaconFile) {
        echo "Beacon sent successfully\n";
        while ($line = fgets($beaconFile)) {
            $trimmed = trim($line);
            if ($trimmed != "")
                echo "$trimmed\n";
        }
        fclose($beaconFile);
        unlink($beaconFilename);
    } else {
        echo "Beacon status not readable\n";
    }
} else {
    echo "Beacon not sent\n";
}
?>

File: LayoutTests/http/tests/blink/sendbeacon/resources/check-beacon.py
#!/usr/bin/env python3

import os
import sys
import tempfile
import time

file = __file__.split(':/cygwin')[-1]
http_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(os.path.dirname(file)))))
sys.path.insert(0, http_root)

from resources.portabilityLayer import get_request


def extensive_strip(text):
    # Sometimes the string ends with \x00
    # and .strip() does not remove it
    text = text.strip()
    while text[-1].encode() == b'\x00':
        text = text[:-1]

    return text


request = get_request()
beacon_filename = os.path.join(tempfile.gettempdir(), 'beacon{}.txt'.format(request.get('name', '')))

max_attempts = 700
retries = int(request.get('retries', max_attempts))
while not os.path.isfile(beacon_filename) and retries != 0:
    time.sleep(0.01)
    retries -= 1

sys.stdout.write(
    'Content-Type: text/plain\r\n'
    'Access-Control-Allow-Origin: *\r\n\r\n'
)

if os.path.isfile(beacon_filename):
    with open(beacon_filename, 'r') as beacon_file:
        sys.stdout.write('Beacon sent successfully\n')
        for line in beacon_file.readlines():
            trimmed = extensive_strip(line)
            if trimmed != '':
                sys.stdout.write('{}\n'.format(trimmed))

    os.remove(beacon_filename)
else:
    sys.stdout.write('Beacon not sent\n')
